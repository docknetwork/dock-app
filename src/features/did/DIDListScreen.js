import {
  Footer,
  Content,
  Button,
  Text,
  Container,
  H1,
  View,
  List,
  ListItem,
  Left,
  Right,
  Icon,
  Row,
  Grid,
  Col,
  Picker,
} from 'native-base';
import React, {useEffect, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {ScreenSpinner} from '../../components/ScreenSpinner';
import {didOperations, didSelectors} from './did-slice';
import styled from 'styled-components/native';
import {Modal} from '../../components/Modal';
import {Colors} from '../../theme/colors';
import {WalletPicker} from '../wallets/UnlockWalletScreen';
import { Dimensions } from 'react-native';

const Divier = styled(View)`
  margin: 20px 0;
  border: 0 solid #ccc;
  border-bottom-width: 1px;
`;

function DetailsValue({label, value}) {
  return (
    <View style={{marginBottom: 12}}>
      <Text style={{marginBottom: 5}}>{label}</Text>
      <Text style={{color: '#666', fontSize: 13}}>{value}</Text>
    </View>
  );
}

function DIDEditModal({items, onClose, onEdit, itemDetails}) {
  const [newController, setNewController] = useState(itemDetails.publicKey[0].controller);

  return (
    <Modal visible={true} onClose={onClose}>
      <Grid>
        <Row
          style={{
            alignItems: 'center',
            height: 70,
            backgroundColor: Colors.darkBlue,
            padding: 10,
          }}>
          <Col style={{width: 50}}>
            <Button transparent onPress={onClose}>
              <Icon name="chevron-back-outline" style={{color: 'white'}} />
              {/* <Text style={{color: '#fff'}}>Delete</Text> */}
            </Button>
          </Col>
          <Col>
            <H1 style={{color: '#fff'}}>Update DID</H1>
          </Col>
        </Row>
        <Row style={{padding: 10, flex: 1}}>
          <Col>
            <DetailsValue label="Id" value={itemDetails.id} />
            {/* <DetailsValue
              label="Controller"
              value={}
            /> */}
            <Text style={{}}>Controller</Text>
            <Picker
              note
              mode="dropdown"
              style={{width: Dimensions.get('screen').width, height: 100}}
              iosIcon={<Icon name="arrow-down" />}
              placeholder="Select a DID"
              placeholderStyle={{color: '#333'}}
              selectedValue={newController}
              onValueChange={(v) => setNewController(v)}>
              {items.map(item => (
                <Picker.Item
                  label={
                    <View style={{paddingTop: 8, paddingLeft: 5}}>
                      <Text style={{color: '#555'}}>{item.id}</Text>
                    </View>
                  }
                  value={item.id}
                  key={item.id}
                />
              ))}
            </Picker>
            <View style={{marginTop: 12}}>
              <Button full disabled={!newController}>
                <Text style={{color: 'white'}}>Confirm Update</Text>
              </Button>
            </View>
          </Col>
        </Row>
      </Grid>
    </Modal>
  );
}

function DIDDetailsModal({onClose, itemDetails, onEdit, onDelete}) {
  return (
    <Modal visible={true} onClose={onClose}>
      <Grid>
        <Row
          style={{
            alignItems: 'center',
            height: 70,
            backgroundColor: Colors.darkBlue,
            padding: 10,
          }}>
          <Col>
            <H1 style={{color: '#fff'}}>DID Details</H1>
          </Col>
          <Col style={{width: 120, flexDirection: 'row'}}>
            <Button transparent onPress={onEdit} style={{marginRight: 8}}>
              <Icon name="pencil" style={{color: 'white'}} />
              {/* <Text style={{color: '#fff'}}>Delete</Text> */}
            </Button>
            <Button transparent onPress={onDelete}>
              <Icon name="trash" style={{color: 'white'}} />
              {/* <Text style={{color: '#fff'}}>Delete</Text> */}
            </Button>
          </Col>
        </Row>
        <Row style={{padding: 10, marginTop: 8}}>
          <Col>
            <DetailsValue label="Id" value={itemDetails.id} />

            <DetailsValue
              label="Authentication"
              value={itemDetails.authentication}
            />

            <DetailsValue
              label="Assertion Method"
              value={itemDetails.assertionMethod}
            />

            <DetailsValue
              label="Controller"
              value={itemDetails.publicKey[0].controller}
            />
            <DetailsValue
              label="Public Key Type"
              value={itemDetails.publicKey[0].type}
            />
            <DetailsValue
              label="Public Key Base58"
              value={itemDetails.publicKey[0].publicKeyBase58}
            />
          </Col>
        </Row>
      </Grid>
    </Modal>
  );
}

export function DIDListScreen({navigation}) {
  const dispatch = useDispatch();
  const items = useSelector(didSelectors.getItems);
  const isLoading = useSelector(didSelectors.getLoading);
  const [itemDetails, setItemDetails] = useState();
  const [showEditModal, setShowEditModal] = useState();

  const handleCreate = () => {
    dispatch(didOperations.createDID());
  };

  const handleDelete = () => {
    dispatch(didOperations.removeDID(itemDetails));
    setItemDetails(null);
  };

  const handleEdit = () => {
    setShowEditModal(true);
  };

  useEffect(() => {
    dispatch(didOperations.fetch());
  }, []);

  return (
    <Container>
      {/* <Header style={{alignItems: 'center', marginTop: 10}}>
        <H1>DIDs</H1>
      </Header> */}
      <Content>
        {isLoading ? (
          <ScreenSpinner />
        ) : (
          <List>
            {(items || []).map(item => (
              <ListItem onPress={() => setItemDetails(item)}>
                <Left>
                  <Text style={{color: '#444', fontSize: 14}}>{item.id}</Text>
                </Left>
                <Right>
                  <Icon name="arrow-forward" />
                </Right>
              </ListItem>
            ))}
          </List>
        )}
        {showEditModal ? (
          <DIDEditModal
            itemDetails={itemDetails}
            onClose={() => setShowEditModal(false)}
            items={items}
          />
        ) : null}
        {!showEditModal && itemDetails ? (
          <DIDDetailsModal
            onClose={() => setItemDetails(null)}
            itemDetails={itemDetails}
            onEdit={handleEdit}
            onDelete={handleDelete}
          />
        ) : null}
      </Content>
      <Footer style={{backgroundColor: '#fff', padding: 8}}>
        <Button full style={{width: '100%'}} primary onPress={handleCreate}>
          <Text style={{color: '#fff'}}>Create DID</Text>
        </Button>
      </Footer>
    </Container>
  );
}
