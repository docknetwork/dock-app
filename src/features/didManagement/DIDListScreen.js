import React from 'react';
import {
  BigButton,
  Box,
  Footer,
  Header,
  IconButton,
  NBox,
  ScreenContainer,
  Theme,
  Typography,
} from '../../design-system';
import {translate} from '../../locales';
import {addTestId} from '../../core/automation-utils';
import PlusCircleWhiteIcon from '../../assets/icons/plus-circle-white.svg';
import {RoundedDIDIcon} from '../../assets/icons';
import {Button, ScrollView} from 'native-base';
import {Ionicons} from '@native-base/icons';
import {Icon, VStack} from 'native-base';
import PlusCircleIcon from '../../assets/icons/plus-circle.svg';
import DocumentDownloadIcon from '../../assets/icons/document-download.svg';
import {navigate} from '../../core/navigation';
import {Routes} from '../../core/routes';

export function DIDListScreen() {
  return (
    <ScreenContainer testID="DIDListScreen" showTabNavigation>
      <Header>
        <Box
          marginLeft={22}
          marginRight={22}
          flexDirection="row"
          alignItems="center">
          <Box flex={1}>
            <Typography fontFamily="Montserrat" fontSize={24} fontWeight="600">
              {translate('app_navigation.did_management')}
            </Typography>
          </Box>
          <Box row>
            <IconButton
              col
              {...addTestId('DIDListScreen_Add_DID')}
              onPress={() => {
                navigate(Routes.DID_MANAGEMENT_NEW_DID);
              }}>
              <PlusCircleWhiteIcon />
            </IconButton>
          </Box>
        </Box>
      </Header>
      <ScrollView marginLeft={26} marginRight={26}>
        <NBox mt={70}>
          <VStack space={7} alignItems="center">
            <RoundedDIDIcon />
            <Typography mt={3} textAlign="center" variant="didDescription">
              {translate('didManagement.did_definition')}
            </Typography>
            <Button
              onPress={null}
              isActive={false}
              endIcon={<Icon as={Ionicons} name="open-outline" size="xs" />}
              variant={'transactionFilter'}
              size={'xs'}>
              <Typography
                variant="transaction-filter"
                color={Theme.colors.inactiveText}>
                {translate('didManagement.learn_more')}
              </Typography>
            </Button>
          </VStack>
        </NBox>
      </ScrollView>
      <Footer marginBottom={114} marginLeft={3} marginRight={3} flex={1}>
        <BigButton
          {...addTestId('CreateNewDID')}
          onPress={null}
          icon={<PlusCircleIcon />}>
          {translate('didManagement.create_new_did')}
        </BigButton>
        <BigButton
          {...addTestId('ImportExistingDIDBtn')}
          onPress={null}
          icon={<DocumentDownloadIcon />}>
          {translate('didManagement.import_existing_did')}
        </BigButton>
      </Footer>
    </ScreenContainer>
  );
}
export function DIDListScreenContainer() {
  return <DIDListScreen />;
}
