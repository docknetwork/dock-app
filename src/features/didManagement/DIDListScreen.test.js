import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {DIDListScreen} from './DIDListScreen';

const mockStore = configureMockStore();

describe('DIDListScreen', () => {
  it('should render correctly ', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(<DIDListScreen />, {
      context: {store: mockStore(initialState)},
    });
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
