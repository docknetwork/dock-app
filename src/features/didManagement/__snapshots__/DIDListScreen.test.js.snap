// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DIDListScreen should render correctly  1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#000",
      "flex": 1,
    }
  }
  testID="DIDListScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <Box
          alignItems="center"
          flexDirection="row"
          marginLeft={22}
          marginRight={22}
        >
          <Box
            flex={1}
          >
            <Styled(Component)
              fontFamily="Montserrat"
              fontSize={24}
              fontWeight="600"
            >
              DIDs
            </Styled(Component)>
          </Box>
          <Box
            row={true}
          >
            <IconButton
              accessibilityLabel="DIDListScreen_Add_DID"
              col={true}
              onPress={[Function]}
              testID="DIDListScreen_Add_DID"
            >
              <SvgFile />
            </IconButton>
          </Box>
        </Box>
      </Styled(View)>
      <ForwardRef
        marginLeft={26}
        marginRight={26}
      >
        <Memo(ForwardRef(Box))
          mt={70}
        >
          <Memo(ForwardRef(VStack))
            alignItems="center"
            space={7}
          >
            <SvgFile />
            <Styled(Component)
              mt={3}
              textAlign="center"
              variant="didDescription"
            >
              Decentralized Identifiers (DIDs) are a new type of identifier that enables a decentralized digital identity and allows you to accept verifiable credentials.
            </Styled(Component)>
            <Memo(ForwardRef(Button))
              endIcon={
                <Memo
                  as={[Function]}
                  name="open-outline"
                  size="xs"
                />
              }
              isActive={false}
              onPress={null}
              size="xs"
              variant="transactionFilter"
            >
              <Styled(Component)
                color="#fff"
                variant="transaction-filter"
              >
                Learn More
              </Styled(Component)>
            </Memo(ForwardRef(Button))>
          </Memo(ForwardRef(VStack))>
        </Memo(ForwardRef(Box))>
      </ForwardRef>
      <Styled(Component)
        flex={1}
        marginBottom={114}
        marginLeft={3}
        marginRight={3}
      >
        <BigButton
          accessibilityLabel="CreateNewDID"
          icon={<SvgFile />}
          onPress={null}
          testID="CreateNewDID"
        >
          Create New DID
        </BigButton>
        <BigButton
          accessibilityLabel="ImportExistingDIDBtn"
          icon={<SvgFile />}
          onPress={null}
          testID="ImportExistingDIDBtn"
        >
          Import existing DID
        </BigButton>
      </Styled(Component)>
    </Memo(ForwardRef(Box))>
    <Memo(ForwardRef(Box))
      flex={1}
      flexGrow={0.11}
    >
      <TabNavigation />
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;
